[metadata] Use bilrost for object store encoding

muhamadazmy opened the pull request:
[metadata] Use bilrost for object store encoding

Summary:
Move away from ciborium and use bilrost instead for
metadata object store

Fixes #3480

pcholakov commented on crates/metadata-providers/src/objstore/version_repository.rs:
```diff
@@ -42,9 +47,44 @@ impl Tag {
+            "binary/bilrost" => Ok(Self::Bilrost),
+            "binary/ciborium" => Ok(Self::Ciborium),
+            // default for backward compatibility
+            "" => Ok(Self::Ciborium),
```
The default is probably unset so we'll just fall back to the `Default` impl, right? Maybe more explicit bail for empty attributes too since we've never set this object attribute before. IDK, probably doesn't hurt.

muhamadazmy commented on crates/metadata-providers/src/objstore/version_repository.rs:
```diff
@@ -42,9 +47,44 @@ impl Tag {
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "binary/bilrost" => Ok(Self::Bilrost),
+            "binary/ciborium" => Ok(Self::Ciborium),
```
Oh, right! thank you for pointing this out.

muhamadazmy commented on crates/metadata-providers/src/objstore/optimistic_store.rs:
```diff
@@ -89,15 +106,27 @@ impl OptimisticLockingMetadataStore {
+                    .map(|_| self.arena.split().freeze())
+                    .map_err(|e| WriteError::Codec(e.into()))
+            }
+            ValueEncoding::Bilrost => {
```
Good catch, thank you. I was wondering what it was for. Thank you for pointing this out.

muhamadazmy commented on crates/metadata-providers/src/objstore/version_repository.rs:
```diff
@@ -42,9 +47,44 @@ impl Tag {
+
+impl FromStr for ValueEncoding {
+    type Err = EncodingError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
```
Thanks for the suggestion! I considered using StorageCodec, but in this case, I don't think it's the right fit. The main reason is that StorageCodec prepends the buffer with the codec identifier, which we don’t need here. We already rely on the `content-encoding` attribute to indicate the encoding.

Also, implementing the StorageEncode/StorageDecode traits cleanly would require adding support for CBOR, which we currently don’t use or plan to support. As for supporting JSON for metadata in the future, that doesn’t seem likely based on our current direction.

So I’d prefer to keep the current approach, unless there is a strong reason that I am missing here :)

pcholakov replied:
Gotcha! Thanks for the detailed answer :-)

AhmedSoliman commented on crates/metadata-providers/src/objstore/optimistic_store.rs:
```diff
@@ -89,15 +118,33 @@ impl OptimisticLockingMetadataStore {
+                };
+
+                salted
+                    .encode(&mut self.arena)
```
Should we reserve() the expected length first from the arena? In other words, how sure are we that the arena is big enough to encode the value?

muhamadazmy replied:
That's a really good catch. Thank you for pointing this out. the arena had enough capacity for the tests, but would fail eventually.

muhamadazmy commented:
Thank you both @pcholakov and @igalshilman for your review. I addressed most of your comments. Would love if you can give this another go and let me know if there are more comments.

AhmedSoliman commented on crates/metadata-providers/src/objstore/optimistic_store.rs:
```diff
@@ -46,14 +50,39 @@ enum OnDiskValue<'a> {
 
 impl OptimisticLockingMetadataStore {
+    // TODO: Switch to bilrost in 1.6.0
+    const DEFAULT_ENCODING: ValueEncoding = ValueEncoding::Cbor;
```
This is redundant, ValueEncoding implements `Default` and it'll be only confusing to have two places with default value set.

pcholakov commented on crates/metadata-providers/src/objstore/optimistic_store.rs:
```diff
@@ -106,24 +151,29 @@ impl OptimisticLockingMetadataStore {
-        // create a random cookie
-        let cookie = random::<u64>();
-        let buf = self.serialize_versioned_value(&value, cookie)?;
+        // create a random salt
+        // The salt is a random value to break CAS ties.
+        // S3 cas working by comparing ETags (hash of the body) and we don't want a case where two writer think they both succeeded the cas (they both CASed the same body)
```
Minor cleanup for your consideration:

```suggestion
        // The salt is a random value to break CAS ties. S3 CAS works by comparing ETags (body hashes).
        // We want to avoid two writers thinking they both succeeded if they happen to write identical bodies.
```

muhamadazmy replied:
Done! :)

pcholakov reviewed (approved):
Thank you! Left a minor comment suggestion.

If you want, you can wait until the "Docker" CI step is complete, and kick off a metadata store Jepsen test at https://github.com/restatedev/jepsen/actions/workflows/jepsen.yml:

Hit "Run workflow:

- put in the PR number without the # `3485` in the PR field
- update the workload string to `set-mds` - so the full input will be: `{ "workloads": "set-mds", "nemeses": "partition-random-node", "duration": "60", "rate": "100", "concurrency": "5n", "testCount": "50" }`

Let it run. Should take ~3h. This doesn't run on main as we don't build with `metadata-api` feature enabled (yet, I believe).
